drivers/watchdog/bflb_wdt.c

```diff
@@ -0,0 +1,297 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/err.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/utsname.h>
+#include <linux/io.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/watchdog.h>
+
+#define BFLB_INT_TICKS_PER_SEC 1024
+#define BFLB_TICK_DIV 256
+#define BFLB_TICKS_PER_SEC ( BFLB_INT_TICKS_PER_SEC / BFLB_TICK_DIV )
+#define BFLB_MAX_SECS ( 65535 / BFLB_TICKS_PER_SEC )
+
+#define BFLB_DEFAULT_TIMEOUT 60
+
+#define BFLB_REG_BASE 0x2000A500
+
+#define BFLB_REG_WFAR 0x9C
+#define BFLB_VAL_WFAR 0xBABA
+
+#define BFLB_REG_WSAR 0xA0
+#define BFLB_VAL_WSAR 0xEB10
+
+#define BFLB_REG_WVR 0x6C
+#define BFLB_REG_WCR 0x98
+#define BFLB_VAL_WCR 0x1
+
+#define BFLB_REG_WMER 0x64
+#define BFLB_VAL_WE 0x1
+#define BFLB_VAL_WRIE 0x2
+
+#define BFLB_REG_TCCR 0x00
+#define BFLB_MASK_CS_WDT 0xF00
+#define BFLB_VAL_CS_1K 0x200
+
+#define BFLB_REG_TCDR 0xbc
+#define BFLB_MASK_TCDR 0xFF000000
+#define BFLB_VAL_TCDR 0xFF000000
+
+#define BFLB_REG_WMR 0x68
+#define BFLB_MASK_WMR 0xffff
+#define BFLB_VAL_WMR 240
+
+
+struct bflb_watchdog_device {
+	struct watchdog_device wdd;
+	struct device *dev;
+	void __iomem *regs;
+};
+
+static inline
+struct bflb_watchdog_device *to_bflb_wdd(struct watchdog_device *wdd)
+{
+	return container_of(wdd, struct bflb_watchdog_device, wdd);
+}
+
+// Access key registers must be written before write
+// operations presumably to prevent accidentally enabling
+// the watchdog and killing the machine.
+static inline int bflb_unlock_watchdog(struct bflb_watchdog_device *bflb_wdd)
+{
+	writew(BFLB_VAL_WFAR, bflb_wdd->regs + BFLB_REG_WFAR);
+	writew(BFLB_VAL_WSAR, bflb_wdd->regs + BFLB_REG_WSAR);
```

Comment:
The peripheral exposes 32-bit registers, so you should always use `writel`, even if some bits are unused.

Reply:
I was mimicking the raw metal bl_mcu_sdk example here, assuming there was an intentional reason to do things this way. I'll change it and test.

Reply:
I did the requested fix in [2165429](https://github.com/arm000/linux-bl808/pull/8/commits/2165429a7e9dd088b2bf67966350adb34abf6004) but then the watchdog doesn't function. I've reverted the fix in [dea6a60](https://github.com/arm000/linux-bl808/pull/8/commits/dea6a6077ccce4a05c8a57043e52344dc0151257) to restore functionality.

I'll continue to look to see if my initial fix is incorrect or there are issues with these security registers.